stat_t layout and size on MIPS32-musl-linux is seemingly wrong

The size of `stat_t` when printed by C program on OpenWRT (mipsel32 platform using the musl libc) is 160. The following D code asserts here in `core.sys.posix.sys.stat`:
```d
        static if (!__USE_FILE_OFFSET64)
            static assert(stat_t.sizeof == 144);
        else
            static assert(stat_t.sizeof == 160);
```

Error:
```
core/sys/posix/sys/stat.d(319): Error: static assert:  `168u == 144u` is false
```

There seems to be multiple things wrong here. First I guess the second assert must be used when using musl libc, since multiple fields in the struct are 64-bit wide (like `off_t`).

Second there're multiple mismatches of the struct compared to [here](https://github.com/bminor/musl/blob/0b86d60badad6a69b37fc06d18b5763fbbf47b58/arch/mips/kstat.h#L1):
```c
//kstat.h
struct kstat {
	unsigned st_dev;
	long __st_padding1[3];
	ino_t st_ino;
	mode_t st_mode;
	nlink_t st_nlink;
	uid_t st_uid;
	gid_t st_gid;
	unsigned st_rdev;
	long __st_padding2[3];
	off_t st_size;
	long st_atime_sec;
	long st_atime_nsec;
	long st_mtime_sec;
	long st_mtime_nsec;
	long st_ctime_sec;
	long st_ctime_nsec;
	blksize_t st_blksize;
	long __st_padding3;
	blkcnt_t st_blocks;
        long __st_padding4[14];
};
``` 
Compare the D code:
```d
        struct stat_t
        {
            c_ulong     st_dev;
            c_long[3]   st_pad1;
            ino_t       st_ino;
            mode_t      st_mode;
            nlink_t     st_nlink;
            uid_t       st_uid;
            gid_t       st_gid;
            c_ulong     st_rdev;
            static if (!__USE_FILE_OFFSET64)
            {
                c_long[2]   st_pad2;
                off_t       st_size;
                c_long      st_pad3;
            }
            else
            {
                c_long[3]   st_pad2;
                off_t       st_size;
            }
            static if (_DEFAULT_SOURCE || _XOPEN_SOURCE >= 700)
            {
                timespec    st_atim;
                timespec    st_mtim;
                timespec    st_ctim;
                extern(D) @safe @property inout pure nothrow
                {
                    ref inout(time_t) st_atime() return { return st_atim.tv_sec; }
                    ref inout(time_t) st_mtime() return { return st_mtim.tv_sec; }
                    ref inout(time_t) st_ctime() return { return st_ctim.tv_sec; }
                }
            }
            else
            {
                time_t      st_atime;
                c_ulong     st_atimensec;
                time_t      st_mtime;
                c_ulong     st_mtimensec;
                time_t      st_ctime;
                c_ulong     st_ctimensec;
            }
            blksize_t   st_blksize;
            static if (!__USE_FILE_OFFSET64)
            {
                blkcnt_t    st_blocks;
            }
            else
            {
                c_long      st_pad4;
                blkcnt_t    st_blocks;
            }
            c_long[14]  st_pad5;
        }
```

The fields `st_atime`, `st_mtime` and `st_ctime` must be of type `c_ulong`, not `time_t` since `time_t` is always 64-bit on musl even on 32-bit systems.

Third, because `__USE_FILE_OFFSET64` is `false`, the struct layout does not match the one in the link. However, If i set `__USE_FILE_OFFSET64` to `true`, the struct grows to approx 180 bytes, so I'm not sure about the correct corrective action needed here.

Probably this should be a new issue report on the LDC repo, but since you ask.

I'm using the following command:
```
CC=mipsel-openwrt-linux-musl-gcc ldc-build-runtime --ninja --dFlags="-mtriple=mipsel-openwrt-linux-musl;-mcpu=mips32;-preview=bitfields" --cFlags="-march=mips32"
```

The output shows:
```
...
core/thread/osthread.d(1589):1:12: warning: used $at (currently $1) without ".set noat"
        sw $23, 0($1)
                  ^
core/thread/osthread.d(1589):1:12: warning: used $at (currently $1) without ".set noat"
        sw $23, 0($1)
                  ^
core/thread/osthread.d(1589):1:12: warning: used $at (currently $1) without ".set noat"
        sw $23, 0($1)
                  ^
ninja: build stopped: subcommand failed.
.: Error: command failed with status 1
```

So I managed to build a static library with:
`CC=mipsel-openwrt-linux-musl-gcc ldc-build-runtime --ninja --dFlags="-mtriple=mipsel-openwrt-linux-musl;-preview=bitfields;-wi;-mcpu=mips32r2"  BUILD_SHARED_LIBS=OFF`

Trying to link with:
`ldc2 -mtriple=mipsel-openwrt-linux-musl -gcc=mipsel-openwrt-linux-musl-gcc test.d -preview=bitfields -L-L/path/to/static/lib`
results in errors about mmap64:
```
io.d:(.text._D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z7ElfFile4openFNbNiPxaJSQDzQDxQDrQDq__TQDqTQDnTQCuVhi1ZQCcZb[_D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z7ElfFile4openFNbNiPxaJSQDzQDxQDrQDq__TQDqTQDnTQCuVhi1ZQCcZb]+0xf8): undefined reference to `mmap64'
/home/ndf/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-13.3.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/13.3.0/../../../../mipsel-openwrt-linux-musl/bin/ld: io.d:(.text._D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z7ElfFile4openFNbNiPxaJSQDzQDxQDrQDq__TQDqTQDnTQCuVhi1ZQCcZb[_D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z7ElfFile4openFNbNiPxaJSQDzQDxQDrQDq__TQDqTQDnTQCuVhi1ZQCcZb]+0x110): undefined reference to `mmap64'
/home/ndf/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-13.3.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/13.3.0/../../../../mipsel-openwrt-linux-musl/bin/ld: /home/ndf/Projects/Dlang/misc/mips/ldc-build-runtime.tmp/lib//libdruntime-ldc.a(io.o): in function `_D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z13NamedSections7opApplyMFNbNiMDFNbNikAxaSQErQEpQEjQEi__TQEiTQEfTQDmVhi1Z16ElfSectionHeaderZiZi':
io.d:(.text._D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z13NamedSections7opApplyMFNbNiMDFNbNikAxaSQErQEpQEjQEi__TQEiTQEfTQDmVhi1Z16ElfSectionHeaderZiZi[_D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z13NamedSections7opApplyMFNbNiMDFNbNikAxaSQErQEpQEjQEi__TQEiTQEfTQDmVhi1Z16ElfSectionHeaderZiZi]+0xb8): undefined reference to `mmap64'
/home/ndf/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-13.3.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/13.3.0/../../../../mipsel-openwrt-linux-musl/bin/ld: io.d:(.text._D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z13NamedSections7opApplyMFNbNiMDFNbNikAxaSQErQEpQEjQEi__TQEiTQEfTQDmVhi1Z16ElfSectionHeaderZiZi[_D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z13NamedSections7opApplyMFNbNiMDFNbNikAxaSQErQEpQEjQEi__TQEiTQEfTQDmVhi1Z16ElfSectionHeaderZiZi]+0xd0): undefined reference to `mmap64'
/home/ndf/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-13.3.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/13.3.0/../../../../mipsel-openwrt-linux-musl/bin/ld: io.d:(.text._D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z13NamedSections7opApplyMFNbNiMDFNbNikAxaSQErQEpQEjQEi__TQEiTQEfTQDmVhi1Z16ElfSectionHeaderZiZi[_D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z13NamedSections7opApplyMFNbNiMDFNbNikAxaSQErQEpQEjQEi__TQEiTQEfTQDmVhi1Z16ElfSectionHeaderZiZi]+0x16c): undefined reference to `mmap64'
/home/ndf/openwrt-sdk/staging_dir/toolchain-mipsel_24kc_gcc-13.3.0_musl/bin/../lib/gcc/mipsel-openwrt-linux-musl/13.3.0/../../../../mipsel-openwrt-linux-musl/bin/ld: /home/ndf/Projects/Dlang/misc/mips/ldc-build-runtime.tmp/lib//libdruntime-ldc.a(io.o):io.d:(.text._D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z13NamedSections7opApplyMFNbNiMDFNbNikAxaSQErQEpQEjQEi__TQEiTQEfTQDmVhi1Z16ElfSectionHeaderZiZi[_D4core8internal3elf2io__T5ElfIOTSQBg3sysQz10Elf32_EhdrTSQCdQxQBu10Elf32_ShdrVhi1Z13NamedSections7opApplyMFNbNiMDFNbNikAxaSQErQEpQEjQEi__TQEiTQEfTQDmVhi1Z16ElfSectionHeaderZiZi]+0x184): more undefined references to `mmap64' follow
collect2: error: ld returned 1 exit status

```

--- a/druntime/src/core/sys/posix/config.d
+++ b/druntime/src/core/sys/posix/config.d
@@ -87,8 +87,8 @@ else version (CRuntime_Musl)
     // Not present in Musl sources
     enum __REDIRECT          = false;
 
-    // Those three are irrelevant for Musl as it always uses 64 bits off_t
-    enum __USE_FILE_OFFSET64 = false;
+    // Always use code paths that are compatible with 64 bits off_t
+    enum __USE_FILE_OFFSET64 = true;
     enum __USE_LARGEFILE     = __USE_FILE_OFFSET64 && !__REDIRECT;
     enum __USE_LARGEFILE64   = __USE_FILE_OFFSET64 && !__REDIRECT;
 

--- a/druntime/src/core/sys/posix/sys/stat.d
+++ b/druntime/src/core/sys/posix/sys/stat.d
@@ -72,29 +72,38 @@ version (linux)
 {
     version (X86)
     {
-        struct stat_t
+        version (CRuntime_Musl)
         {
-            dev_t       st_dev;
-            ushort      __pad1;
-            static if (!__USE_FILE_OFFSET64)
+            struct stat_t
             {
+                dev_t       st_dev;
+                ushort      __pad1;
+                static if (!__USE_FILE_OFFSET64)
+                {
+                    ino_t       st_ino;
+                }
+                else
+                {
+                    uint        __st_ino;
+                }
+                mode_t      st_mode;
+                nlink_t     st_nlink;
+                uid_t       st_uid;
+                gid_t       st_gid;
+                dev_t       st_rdev;
+                ushort      __pad2;
+                off_t       st_size;
+                blksize_t   st_blksize;
+                blkcnt_t    st_blocks;
+                private struct __timespec32
+                {
+                    c_long tv_sec;
+                    c_long tv_nsec;
+                }
+                __timespec32 __st_atim32;
+                __timespec32 __st_mtim32;
+                __timespec32 __st_ctim32;
                 ino_t       st_ino;
-            }
-            else
-            {
-                uint        __st_ino;
-            }
-            mode_t      st_mode;
-            nlink_t     st_nlink;
-            uid_t       st_uid;
-            gid_t       st_gid;
-            dev_t       st_rdev;
-            ushort      __pad2;
-            off_t       st_size;
-            blksize_t   st_blksize;
-            blkcnt_t    st_blocks;
-            static if (_DEFAULT_SOURCE || _XOPEN_SOURCE >= 700)
-            {
                 timespec    st_atim;
                 timespec    st_mtim;
                 timespec    st_ctim;
@@ -105,23 +114,61 @@ version (linux)
                     ref inout(time_t) st_ctime() return { return st_ctim.tv_sec; }
                 }
             }
-            else
+            static assert(stat_t.sizeof == 144);
+        }
+        else
+        {
+            struct stat_t
             {
-                time_t      st_atime;
-                ulong_t     st_atimensec;
-                time_t      st_mtime;
-                ulong_t     st_mtimensec;
-                time_t      st_ctime;
-                ulong_t     st_ctimensec;
-            }
-            static if (__USE_FILE_OFFSET64)
-            {
-                ino_t       st_ino;
-            }
-            else
-            {
-                c_ulong     __unused4;
-                c_ulong     __unused5;
+                dev_t       st_dev;
+                ushort      __pad1;
+                static if (!__USE_FILE_OFFSET64)
+                {
+                    ino_t       st_ino;
+                }
+                else
+                {
+                    uint        __st_ino;
+                }
+                mode_t      st_mode;
+                nlink_t     st_nlink;
+                uid_t       st_uid;
+                gid_t       st_gid;
+                dev_t       st_rdev;
+                ushort      __pad2;
+                off_t       st_size;
+                blksize_t   st_blksize;
+                blkcnt_t    st_blocks;
+                static if (_DEFAULT_SOURCE || _XOPEN_SOURCE >= 700)
+                {
+                    timespec    st_atim;
+                    timespec    st_mtim;
+                    timespec    st_ctim;
+                    extern(D) @safe @property inout pure nothrow
+                    {
+                        ref inout(time_t) st_atime() return { return st_atim.tv_sec; }
+                        ref inout(time_t) st_mtime() return { return st_mtim.tv_sec; }
+                        ref inout(time_t) st_ctime() return { return st_ctim.tv_sec; }
+                    }
+                }
+                else
+                {
+                    time_t      st_atime;
+                    ulong_t     st_atimensec;
+                    time_t      st_mtime;
+                    ulong_t     st_mtimensec;
+                    time_t      st_ctime;
+                    ulong_t     st_ctimensec;
+                }
+                static if (__USE_FILE_OFFSET64)
+                {
+                    ino_t       st_ino;
+                }
+                else
+                {
+                    c_ulong     __unused4;
+                    c_ulong     __unused5;
+                }
             }
         }
     }
@@ -261,37 +308,31 @@ version (linux)
     }
     else version (MIPS_O32)
     {
-        struct stat_t
+        version (CRuntime_Musl)
         {
-            version (CRuntime_Musl)
+            struct stat_t
             {
                 dev_t       st_dev;
-                c_long[2]   st_pad1;
-            }
-            else
-            {
-                c_ulong     st_dev;
-                c_long[3]   st_pad1;
-            }
-            ino_t       st_ino;
-            mode_t      st_mode;
-            nlink_t     st_nlink;
-            uid_t       st_uid;
-            gid_t       st_gid;
-            c_ulong     st_rdev;
-            static if (!__USE_FILE_OFFSET64)
-            {
-                c_long[2]   st_pad2;
+                c_long[2]   __pad1;
+                ino_t       st_ino;
+                mode_t      st_mode;
+                nlink_t     st_nlink;
+                uid_t       st_uid;
+                gid_t       st_gid;
+                dev_t       st_rdev;
+                c_long[2]   __pad2;
                 off_t       st_size;
-                c_long      st_pad3;
-            }
-            else
-            {
-                c_long[3]   st_pad2;
-                off_t       st_size;
-            }
-            static if (_DEFAULT_SOURCE || _XOPEN_SOURCE >= 700)
-            {
+                private struct __timespec32
+                {
+                    c_long tv_sec;
+                    c_long tv_nsec;
+                }
+                __timespec32 __st_atim32;
+                __timespec32 __st_mtim32;
+                __timespec32 __st_ctim32;
+                blksize_t   st_blksize;
+                c_long      __pad3;
+                blkcnt_t    st_blocks;
                 timespec    st_atim;
                 timespec    st_mtim;
                 timespec    st_ctim;
@@ -301,27 +342,65 @@ version (linux)
                     ref inout(time_t) st_mtime() return { return st_mtim.tv_sec; }
                     ref inout(time_t) st_ctime() return { return st_ctim.tv_sec; }
                 }
+                c_long[2]   __pad4;
             }
-            else
+        }
+        else
+        {
+            struct stat_t
             {
-                time_t      st_atime;
-                c_ulong     st_atimensec;
-                time_t      st_mtime;
-                c_ulong     st_mtimensec;
-                time_t      st_ctime;
-                c_ulong     st_ctimensec;
+                c_ulong     st_dev;
+                c_long[3]   st_pad1;
+                ino_t       st_ino;
+                mode_t      st_mode;
+                nlink_t     st_nlink;
+                uid_t       st_uid;
+                gid_t       st_gid;
+                c_ulong     st_rdev;
+                static if (!__USE_FILE_OFFSET64)
+                {
+                    c_long[2]   st_pad2;
+                    off_t       st_size;
+                    c_long      st_pad3;
+                }
+                else
+                {
+                    c_long[3]   st_pad2;
+                    off_t       st_size;
+                }
+                static if (_DEFAULT_SOURCE || _XOPEN_SOURCE >= 700)
+                {
+                    timespec    st_atim;
+                    timespec    st_mtim;
+                    timespec    st_ctim;
+                    extern(D) @safe @property inout pure nothrow
+                    {
+                        ref inout(time_t) st_atime() return { return st_atim.tv_sec; }
+                        ref inout(time_t) st_mtime() return { return st_mtim.tv_sec; }
+                        ref inout(time_t) st_ctime() return { return st_ctim.tv_sec; }
+                    }
+                }
+                else
+                {
+                    time_t      st_atime;
+                    c_ulong     st_atimensec;
+                    time_t      st_mtime;
+                    c_ulong     st_mtimensec;
+                    time_t      st_ctime;
+                    c_ulong     st_ctimensec;
+                }
+                blksize_t   st_blksize;
+                static if (!__USE_FILE_OFFSET64)
+                {
+                    blkcnt_t    st_blocks;
+                }
+                else
+                {
+                    c_long      st_pad4;
+                    blkcnt_t    st_blocks;
+                }
+                c_long[14]  st_pad6;
             }
-            blksize_t   st_blksize;
-            static if (!__USE_FILE_OFFSET64)
-            {
-                blkcnt_t    st_blocks;
-            }
-            else
-            {
-                c_long      st_pad4;
-                blkcnt_t    st_blocks;
-            }
-            c_long[14]  st_pad5;
         }
         static if (!__USE_FILE_OFFSET64)
             static assert(stat_t.sizeof == 144);
@@ -572,67 +651,34 @@ version (linux)
     }
     else version (ARM)
     {
-        private
+        version (CRuntime_Musl)
         {
-            alias __dev_t = ulong;
-            alias __ino_t = c_ulong;
-            alias __ino64_t = ulong;
-            alias __mode_t = uint;
-            alias __nlink_t = size_t;
-            alias __uid_t = uint;
-            alias __gid_t = uint;
-            alias __off_t = c_long;
-            alias __off64_t = long;
-            alias __blksize_t = c_long;
-            alias __blkcnt_t = c_long;
-            alias __blkcnt64_t = long;
-            alias __timespec = timespec;
-            alias __time_t = time_t;
-        }
-        struct stat_t
-        {
-            __dev_t st_dev;
-            ushort __pad1;
-
-            static if (!__USE_FILE_OFFSET64)
+            struct stat_t
             {
-                __ino_t st_ino;
-            }
-            else
-            {
-                __ino_t __st_ino;
-            }
-            __mode_t st_mode;
-            __nlink_t st_nlink;
-            __uid_t st_uid;
-            __gid_t st_gid;
-            __dev_t st_rdev;
-            ushort __pad2;
-
-            static if (!__USE_FILE_OFFSET64)
-            {
-                __off_t st_size;
-            }
-            else
-            {
-                __off64_t st_size;
-            }
-            __blksize_t st_blksize;
-
-            static if (!__USE_FILE_OFFSET64)
-            {
-                __blkcnt_t st_blocks;
-            }
-            else
-            {
-                __blkcnt64_t st_blocks;
-            }
-
-            static if ( _DEFAULT_SOURCE || _XOPEN_SOURCE >= 700)
-            {
-                __timespec st_atim;
-                __timespec st_mtim;
-                __timespec st_ctim;
+                dev_t st_dev;
+                ushort __pad1;
+                c_long __st_ino;
+                mode_t st_mode;
+                nlink_t st_nlink;
+                uid_t st_uid;
+                gid_t st_gid;
+                dev_t st_rdev;
+                ushort __pad2;
+                off_t st_size;
+                blksize_t st_blksize;
+                blkcnt_t st_blocks;
+                private struct __timespec32
+                {
+                    c_long tv_sec;
+                    c_long tv_nsec;
+                }
+                __timespec32 __st_atim32;
+                __timespec32 __st_mtim32;
+                __timespec32 __st_ctim32;
+                ino_t st_ino;
+                timespec st_atim;
+                timespec st_mtim;
+                timespec st_ctim;
                 extern(D) @safe @property inout pure nothrow
                 {
                     ref inout(time_t) st_atime() return { return st_atim.tv_sec; }
@@ -640,30 +686,103 @@ version (linux)
                     ref inout(time_t) st_ctime() return { return st_ctim.tv_sec; }
                 }
             }
-            else
-            {
-                __time_t st_atime;
-                c_ulong st_atimensec;
-                __time_t st_mtime;
-                c_ulong st_mtimensec;
-                __time_t st_ctime;
-                c_ulong st_ctimensec;
-            }
-
-            static if (!__USE_FILE_OFFSET64)
-            {
-                c_ulong __unused4;
-                c_ulong __unused5;
-            }
-            else
-            {
-                __ino64_t st_ino;
-            }
+            static assert(stat_t.sizeof == 152);
         }
-        static if (__USE_FILE_OFFSET64)
-            static assert(stat_t.sizeof == 104);
         else
-            static assert(stat_t.sizeof == 88);
+        {
+            private
+            {
+                alias __dev_t = ulong;
+                alias __ino_t = c_ulong;
+                alias __ino64_t = ulong;
+                alias __mode_t = uint;
+                alias __nlink_t = size_t;
+                alias __uid_t = uint;
+                alias __gid_t = uint;
+                alias __off_t = c_long;
+                alias __off64_t = long;
+                alias __blksize_t = c_long;
+                alias __blkcnt_t = c_long;
+                alias __blkcnt64_t = long;
+                alias __timespec = timespec;
+                alias __time_t = time_t;
+            }
+            struct stat_t
+            {
+                __dev_t st_dev;
+                ushort __pad1;
+
+                static if (!__USE_FILE_OFFSET64)
+                {
+                    __ino_t st_ino;
+                }
+                else
+                {
+                    __ino_t __st_ino;
+                }
+                __mode_t st_mode;
+                __nlink_t st_nlink;
+                __uid_t st_uid;
+                __gid_t st_gid;
+                __dev_t st_rdev;
+                ushort __pad2;
+
+                static if (!__USE_FILE_OFFSET64)
+                {
+                    __off_t st_size;
+                }
+                else
+                {
+                    __off64_t st_size;
+                }
+                __blksize_t st_blksize;
+
+                static if (!__USE_FILE_OFFSET64)
+                {
+                    __blkcnt_t st_blocks;
+                }
+                else
+                {
+                    __blkcnt64_t st_blocks;
+                }
+
+                static if ( _DEFAULT_SOURCE || _XOPEN_SOURCE >= 700)
+                {
+                    __timespec st_atim;
+                    __timespec st_mtim;
+                    __timespec st_ctim;
+                    extern(D) @safe @property inout pure nothrow
+                    {
+                        ref inout(time_t) st_atime() return { return st_atim.tv_sec; }
+                        ref inout(time_t) st_mtime() return { return st_mtim.tv_sec; }
+                        ref inout(time_t) st_ctime() return { return st_ctim.tv_sec; }
+                    }
+                }
+                else
+                {
+                    __time_t st_atime;
+                    c_ulong st_atimensec;
+                    __time_t st_mtime;
+                    c_ulong st_mtimensec;
+                    __time_t st_ctime;
+                    c_ulong st_ctimensec;
+                }
+
+                static if (!__USE_FILE_OFFSET64)
+                {
+                    c_ulong __unused4;
+                    c_ulong __unused5;
+                }
+                else
+                {
+                    __ino64_t st_ino;
+                }
+            }
+	          static if (__USE_FILE_OFFSET64)
+	              static assert(stat_t.sizeof == 104);
+	          else
+	              static assert(stat_t.sizeof == 88);
+        }
     }
     else version (AArch64)
     {

--- a/druntime/src/core/sys/posix/sys/types.d
+++ b/druntime/src/core/sys/posix/sys/types.d
@@ -124,6 +124,7 @@ version (linux)
 
     version (CRuntime_Musl)
     {
+        static assert(off_t.sizeof == 8);
         /**
          * Musl versions before v1.2.0 (up to v1.1.24) had different
          * definitions for `time_t` for 32 bits.

--- a/druntime/src/core/sys/posix/time.d
+++ b/druntime/src/core/sys/posix/time.d
@@ -16,6 +16,7 @@
 module core.sys.posix.time;
 
 import core.sys.posix.config;
+import core.sys.posix.endian;
 public import core.stdc.time;
 public import core.sys.posix.sys.types;
 public import core.sys.posix.signal; // for sigevent
@@ -201,7 +202,11 @@ version (linux)
     struct timespec
     {
         time_t  tv_sec;
+        version (CRuntime_Musl)
+            int : 8 * (time_t.sizeof - c_long.sizeof) * (BYTE_ORDER == BIG_ENDIAN);
         c_long  tv_nsec;
+        version (CRuntime_Musl)
+            int : 8 * (time_t.sizeof - c_long.sizeof) * (BYTE_ORDER != BIG_ENDIAN);
     }
 }
 else version (Darwin)
@@ -461,6 +466,8 @@ else version (CRuntime_Bionic)
 }
 else version (CRuntime_Musl)
 {
+    static assert(timespec.sizeof == 16);
+
     alias int clockid_t;
     alias void* timer_t;
 

